### PR TITLE
Improve efficiency and fix bugs (device and dtype) for Gaussian backend

### PR DESCRIPTION
**Pull request by Jooyuza**

### CODE REVIEW

**Comment by sansiro77** [comment]
It seems to be a bug related to `vmap` + slice assignment + cuda.

For example,
```python
x = torch.randn(2, 4, 4, device='cuda:0')
def f(x):
    x[0:2, 0:2] = 1
    return x
torch.vmap(f)(x)
```
is OK.

However,
```python
x = torch.randn(2, 4, 4, device='cuda:0')
def f(x):
    x[[0,1], [0,1]] = 1
    return x
torch.vmap(f)(x)
```
reports error.

Try something like:
```python
mat = torch.randn(3, 3, device='cuda:0')
x = mat.new_ones(1)
torch.block_diag(*([x] * 3))
```

**Comment by sansiro77** [comment]
Try
```python
mat = torch.randn(3, 3, device='cuda:0')
x = mat.new_ones(3)
torch.diag(x)
```
to improve efficiency.

File: src/deepquantum/photonic/circuit.py [hunk 1]
```diff
@@ -490,7 +490,7 @@ def _forward_gaussian_prob_helper(self, cov, mean, even_basis, odd_basis, basis,
                 probs_i = torch.cat(probs_half)
             else:
                 probs_half = torch.cat(probs_half)
-                probs_i = torch.cat([probs_half.squeeze(), torch.zeros(len(torch.cat(odd_basis)))])
+                probs_i = torch.cat([probs_half.squeeze(), torch.zeros(len(torch.cat(odd_basis)), device=probs_half.device)])
```
Comment (sansiro77) [inline src/deepquantum/photonic/circuit.py]: 这行太长了，torch.zeros前面换行

**Comment by Jooyuza** [comment]
image

**Review by sansiro77 — APPROVED** [review]
Good job!

**Pull request merged**